File: LeetCode Solutions/January 2023 LeetCoding Challenge/Longest Path With Different Adjacent Characters.java
class Solution {
    int result;
    public int longestPath(int[] parent, String s) {
        result = 0;
        ArrayList<Integer>[] children = new ArrayList[parent.length];
        for (int i=0; i<parent.length; i++){
            children[i] = new ArrayList<>();
        }
        for (int i=1; i<parent.length; i++){
            children[parent[i]].add(i);
        }
        helper(children, s, 0);
        return result;
    }

    private int helper(ArrayList<Integer>[] children, String s, int i) {
        PriorityQueue<Integer> queue = new PriorityQueue<>();
        for (int j : children[i]){
            int current = helper(children, s, j);
            if (s.charAt(j) != s.charAt(i)){
                queue.offer(-current);
            }
        }
        int big1 = queue.isEmpty() ? 0 : -queue.poll();
        int big2 = queue.isEmpty() ? 0 : -queue.poll();
        result = Math.max(result, big1 + big2 + 1);
        return big1 + 1;
    }
}

File: LeetCode Solutions/January 2023 LeetCoding Challenge/Longest Path With Different Adjacent Characters.cpp
class Solution {
public:
    int longestPath(vector<int>& parent, string s) {
        int n = s.size(), result = 0;
        vector<vector<int>> children(n, vector<int>());
        for (int i=1; i<n; ++i){
            children[parent[i]].push_back(i);
        }
        helper(children, s, result, 0);
        return result;
    }

    int helper(vector<vector<int>>& children, string& s, int& result, int i) {
        int big1 = 0, big2 = 0;
        for (int& j : children[i]){
            int current = helper(children, s, result, j);
            if (s[i] == s[j]){
                continue;
            }
            if (current > big2){
                big2 = current;
            }
            if (big2 > big1){
                swap(big1, big2);
            }
        }
        result = max(result, big1 + big2 + 1);
        return big1 + 1;
    }
};